packages/bsky/src/notifications.ts

```diff
@@ -0,0 +1,311 @@
+      // if user has no tokens or the post attr cannot be found, skip
+      if (!userTokens || userTokens.length === 0) {
+        continue
+      }
+      if (!attr) {
+        logger.warn(
+          {
+            userDid,
+            notif,
+          },
+          'No notification display attributes found for this notification. Either profile or post data for this notification is missing.',
+        )
+        continue
+      }
+      const { title, body } = attr
+
+      for (const t of userTokens) {
+        const { appId, platform, token } = t
+        if (platform === 'ios' || platform === 'android') {
+          notifsToSend.push({
+            tokens: [token],
+            platform: platform === 'ios' ? 1 : 2,
+            title: title,
+            message: body,
+            topic: appId,
+            data: {
+              reason: notif.reason,
+              recordUri: notif.recordUri,
+              recordCid: notif.recordCid,
+            },
+          })
+        } else {
+          // @TODO: Handle web notifs
+        }
+      }
+    }
+    return notifsToSend
+  }
+
+  /**
+   * The function `addNotificationsToQueue` adds push notifications to a queue, taking into account rate
+   * limiting and batching the notifications for efficient processing.
+   * @param {PushNotification[]} notifs - An array of PushNotification objects. Each PushNotification
+   * object has a "tokens" property which is an array of tokens.
+   * @returns void
+   */
+  async addNotificationsToQueue(notifs: PushNotification[]) {
+    if (!this.redis) {
+      throw new Error('Redis not defined in NotificationServer')
+    }
+    for (const notif of notifs) {
+      const { tokens } = notif
+      for (const token of tokens) {
+        // RATE LIMITING
+        // Get rate limit data for user
+        const key = `notification_rate_limit:${token}`
+        const now = Date.now()
+        const twentyMinutesAgo = now - 20 * 60 * 1000
+        // Remove timestamps outside of 20 minutes window
+        await this.redis.zremrangebyscore(
+          key,
+          Number.NEGATIVE_INFINITY,
+          twentyMinutesAgo,
+        )
+        const currentUserRateCount = await this.redis.zcount(
+          key,
+          twentyMinutesAgo,
+          now,
+        )
+        // If rate limit reached, skip adding notification to be sent
+        if (currentUserRateCount >= 20) {
+          return
+        }
+        // Add timestamp to user's rate limit data
+        await this.redis.zadd(key, now, now)
+        // Set expiration for rate limit data to 25 minutes just in case
+        await this.redis.expire(key, 25 * 60)
+
+        // BATCHING
+        // Add to batch
+        this.notificationBatch.push(notif)
+        // If batch size is 20, add task to background queue
+        if (this.notificationBatch.length >= this.notificationBatchSize) {
+          try {
+            this.backgroundQueue.add(async () => {
```

Comment:
I think we need this `await` a. in order to catch any error below, and b. in order to maintain `this.notificationBatch`:
```suggestion
            await this.backgroundQueue.add(async () => {
```

Is the purpose of using the background queue here to limit concurrency of sending notifications, or some other reason?

Reply:
I think we want to put the entire notification flow on the background queue (out in the indexing service) so that it's non-blocking on write

Reply:
similar to labels

Reply:
This part doesn't have to be in the background queue, as the resource intensive portion `prepareNotifs` is already in the background queue.

I just put it in the background queue because I thought we wanted to limit concurrency, as you said, as much as possible. But I do not think it is required because the function is already called in a backgroundQueue.